develop/items/custom-data-components.md

```diff
@@ -0,0 +1,272 @@
+
+```java
+stack.set(ModComponents.CLICK_COUNT_COMPONENT, newValue);
+```
+
+This takes our component type and the value we want to set it to. In this case it will be our new click count. This method also returns the old value of the component (if one is present) which may be useful in some situations. For example:
+
+```java
+int oldValue = stack.set(ModComponents.CLICK_COUNT_COMPONENT, newValue);
+```
+
+Let's set up a new `use()` method to read the old click count, increase it by one, and then set the updated click count.
+
+@[code transcludeWith=::2](@/reference/latest/src/main/java/com/example/docs/item/custom/CounterItem.java)
+
+Now try starting the game and right-clicking with the Counter item in your hand. If you open up your inventory and look at the item again you should see that the usage number has gone up by the amount of times you've clicked it.
+
+![Tooltip showing "Used 8 times"](/assets/develop/items/custom_component_3.png)
+
+## Removing Component Value {#removing-component-value}
+
+You can also remove a component from your `ItemStack` if it is no longer needed. This is done by using the `remove()` method, which takes in your component type.
+
+```java
+stack.remove(ModComponents.CLICK_COUNT_COMPONENT);
+```
+
+This method also returns the value of the component before being removed, so you can also use it as follows:
+
+```java
+int oldCount = stack.remove(ModComponents.CLICK_COUNT_COMPONENT);
+```
+
+## Advanced Data Components {#advanced-data-components}
+
+You may need to store multiple attributes in a single component. As a vanilla example, the `minecraft:food` component stores several values related to food, such as `nutrition`, `saturation`, `eat_seconds` and more. In this guide we'll refer to them as "composite" components.
+
+For composite components, you must create a `record` class to store the data. This is the type we'll register in our component type and what we'll read and write when interacting with an `ItemStack`. Start by making a new record class in the `component` package we made earlier.
+
+```java
+public record MyCustomComponent() {
+
+}
+```
+
+Notice that there's a set of brackets after the class name. This is where we define the list of properties we want our component to have. Let's add a float and a boolean called `temperature` and `burnt` respectively.
+
+@[code transcludeWith=::1](@/reference/latest/src/main/java/com/example/docs/component/MyCustomComponent.java)
+
+Since we are defining a custom data structure there won't be a pre-existing `Codec` for our use case like with the [basic component](#basic-data-components). This means we're going to have to construct our own codec. Let's define one in our record class using a `RecordCodecBuilder` which we can reference once we register the component. For more details on using a `RecordCodecBuilder` you can refer to [this section of the Codecs page](../codecs#merging-codecs-for-record-like-classes).
```

Comment:
```suggestion
Since we are defining a custom data structure, there won't be a pre-existing `Codec` for our use case like with the [basic component](#basic-data-components). This means we're going to have to construct our own codec. Let's define one in our record class using a `RecordCodecBuilder` which we can reference once we register the component. For more details on using a `RecordCodecBuilder` you can refer to [this section of the Codecs page](../codecs#merging-codecs-for-record-like-classes).
```